File: apps/api/aqapi/quest/dao/quest_dao.py
from typing import List, Optional
from sqlalchemy.ext.asyncio import AsyncSession
from aqapi.quest.entity.quests_entity import QuestsEntity
from aqapi.core.dao.base_dao import BaseDao
from aqapi.core.config.redis_config import redis_client
from aqapi.core.cache.redis_cacher import RedisCacher
cacher = RedisCacher(redis_client)

class QuestDao(BaseDao):
    """クエストDAOクラス"""

    def __init__(self, session: AsyncSession):
        super().__init__(session)

    @property
    def entity_class(self) -> type[QuestsEntity]:
        return QuestsEntity
    
    @cacher.cache("quests:all")
    async def fetch_all(self) -> List[QuestsEntity]:
        return await super().fetch_all()

    @cacher.cache("quests:{id}")
    async def fetch_by_id(self, id: int) -> Optional[QuestsEntity]:
        return await super().fetch_by_id(id)

    @cacher.evict("quests:all")
    async def insert(self, entity: QuestsEntity) -> int:
        return await super().insert(entity)

    @cacher.evict("quests:all", "quests:{entity.id}")
    async def update(self, entity: QuestsEntity) -> None:
        await super().update(entity)

    @cacher.evict("quests:all", "quests:{id}")
    async def delete(self, id: int) -> None:
        await super().delete(id)

File: packages/backend/src/features/quest/dao/questDao.ts
import { EntityManager } from 'typeorm';
import { BaseDao } from '@backend/core/dao/baseDao';
import { QuestEntity } from '../entity/questEntity';
import { cache, evict } from '@backend/core/cache/redisCache';

/**
 * クエストDAOクラス
 */
export class QuestDao extends BaseDao<QuestEntity> {
  constructor(session: EntityManager) {
    super(session);
  }

  protected get entityClass(): new () => QuestEntity {
    return QuestEntity;
  }

  /**
   * 全てのクエストを取得（キャッシュ付き）
   */
  @cache("quests:all")
  async fetchAllWithCache(): Promise<QuestEntity[]> {
    return await super.fetchAll();
  }

  /**
   * IDでクエストを取得（キャッシュ付き）
   */
  @cache("quests:{id}")
  async fetchByIdWithCache(id: number): Promise<QuestEntity | null> {
    return await super.fetchById(id);
  }

  /**
   * クエストを作成（キャッシュクリア付き）
   */
  @evict("quests:all")
  async insertWithCache(entity: QuestEntity): Promise<number> {
    return await super.insert(entity);
  }

  /**
   * クエストを更新（キャッシュクリア付き）
   */
  @evict("quests:all", "quests:{entity.id}")
  async updateWithCache(entity: QuestEntity): Promise<void> {
    await super.update(entity);
  }

  /**
   * クエストを削除（キャッシュクリア付き）
   */
  @evict("quests:all", "quests:{id}")
  async deleteWithCache(id: number): Promise<void> {
    await super.delete(id);
  }
}
